text/0000-inherent-trait-impl.md

```diff
@@ -0,0 +1,152 @@
+- Feature Name: inherent-trait-impl
+- Start Date: 2018-03-27
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Provides a mechanism to declare "inherent traits" for a type defined in the same crate. Methods on these traits are callable on instances of the specified type without needing to import the trait.
+
+# Motivation
+[motivation]: #motivation
+
+There are two similar cases where this is valuable:
+
+- Frequently used traits.
+
+  Sometimes getting the right abstractions require breaking up a type's implementation into many traits, with only a few methods per
+  trait. Every use of such a type results in a large number of imports to ensure the correct traits are in scope. If such a type is used
+  frequently, then this burden quickly becomes a pain point for users of the API,
+  especially if users do not care about writing generic code over traits.
+
+- Mapping object-oriented APIs.
+
+  When mapping these APIs to rust, base classes are usually mapped to traits: methods on those base classes will need to be callable on any
+  derived type. This is sub-optimal because to use a class method a user must now know which class in the hierachy defined that
+  method, so that they can import and use the corresponding trait. This knowledge is not required when using the same API from an
+  object-oriented language.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The feature is implemented using a new attribute which can be applied to trait
+`impl` blocks:
+
+```rust
+pub trait Bar {
+    fn bar(&self);
+}
+
+pub trait ExtBar: Bar {
+    fn ext_bar(&self);
+}
+
+impl<T: Bar> ExtBar for T {
+    fn ext_bar(&self) { }
+}
+
+pub struct Foo;
+
+#[inherent]
+impl Bar for Foo {
+    fn bar(&self) { println!("foo::bar"); }
+}
+
+// works thanks to specialization
+#[inherent]
+impl ExtBar for Foo { }
+
+impl Foo {
+    fn foo(&self) { println!("foo::foo"); }
+}
+```
+
+The methods `bar` and `ext_bar` are now callable on any instance of `Foo`,
+regardless of whether the `Bar` and `ExtBar` traits are currently in scope,
+or even whether the `Bar` trait is publically visible. In other words if `Bar`
+and `ExtBar` defined in one crate and `Foo` and another, user of `Foo` will be
+able to explicitly depend only on the crate which defines `Foo` and still use
+the inherent traits methods.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
```

Comment:
This should be fleshed out, including:

 - What impl blocks it can be applied to (can the trait be foreign?)
 - Whether it works with `#[fundamental]`
 - Does it error if you have an inherent method with the same name?

Reply:
I'd think it should error if you have an inherent method with the same name, like if you had two inherent methods with the same name.

Reply:
>What impl blocks it can be applied to (can the trait be foreign?)

Yes, it can. One of the main use-cases for inherent trait implementations is implementation of crates defined in external crates.

>Whether it works with #[fundamental]

I am not sure why it should not, though I can't say I fully understand `#[fundamental]` semantics.

>Does it error if you have an inherent method with the same name?

Initially I though it should error, but as I wrote below probably it may be better to issue a warning and shadow inherent trait methods by true inherent methods.

I will try to update the text based on your review! Thanks!